gem/gem.py

```diff
@@ -782,6 +783,79 @@ def __new__(cls, i, j):
         return self
 
 
+# TODO: inverse should be different from comp tensor
+# in the sense that it does not take a local to a global view of a matrix
+# but instead eats a global view and spits out another global view
+# it does not transformation between free indices and shape so you would think it does not need a multiindex
+# but the multiindex is necessary to feed in the global view of the tensor
+# it can still have free_indices inherited by the  aggregate (?)
+class Inverse(Node):
+    __slots__ = ('children', 'multiindex', 'shape')
+    __back__ = ('multiindex',)
+
+    def __new__(cls, aggregate, multiindex):
+        assert multiindex
+
+        # Set index extents from shape
+        for index, extent in zip(multiindex, aggregate.shape):
+            assert isinstance(index, Index)
+            index.set_extent(extent)
+
+        self = super(Inverse, cls).__new__(cls)
+        self.children = (aggregate,)
+        self.shape = aggregate.shape
+        self.multiindex = multiindex
+        self.free_indices = aggregate.free_indices
+
+        return self
+
+
+# NOTE: Slate does not support determinants yet
+# determinant taks in a global view and spits out a scalar
+# similar to an Indexed Node: takes a object of some shape
+# and turns into something scalar shaped
+class Determinant(Scalar):
+    __slots__ = ('children', )
+
+    def __new__(cls, aggregate):
+        # det of scalar valued constant is aggregate itself
+        if isinstance(aggregate, Constant):
+            return aggregate
+
+        # tensor must be square
+        assert not aggregate.shape
+        assert all(aggregate.shape) == aggregate.shape[0], "Tensor must be square"
```

Comment:
That is not how `all` works, you probably meant `set`. Hint: this asserts that if no dimensions are 0, then the first dimension must be 1, if any dimension is zero, then the first dimension must be zero... What?...